File: src/main/resources/static/js/jaffolding/core/Router.js
/**
 * Simple client-side router for the Jaffolding framework.
 * JavaScript implementation that mirrors the Java API.
 */
(function() {
  class Router {
    /**
     * Creates a new Router instance.
     * @param {HTMLElement} container - The container element for rendering routes
     */
    constructor(container) {
      this.container = container;
      this.routes = {};
      this.currentPath = null;
      
      // Listen for popstate events (back/forward navigation)
      window.addEventListener('popstate', () => {
        this.navigateTo(this.getPath());
      });
    }

    /**
     * Adds a route to the router.
     * @param {string} path - The route path
     * @param {Function} componentFactory - Function that returns a Component for the route
     * @returns {Router} - The router instance for chaining
     */
    addRoute(path, componentFactory) {
      this.routes[path] = componentFactory;
      return this;
    }

    /**
     * Starts the router.
     */
    start() {
      // Handle initial route
      this.navigateTo(this.getPath());
      
      // Intercept link clicks for client-side routing
      document.addEventListener('click', (event) => {
        const target = event.target.closest('a');
        if (target && 
            target.hasAttribute('href') && 
            !target.getAttribute('href').startsWith('http') &&
            !target.getAttribute('href').startsWith('#')) {
          
          event.preventDefault();
          this.navigateTo(target.getAttribute('href'));
        }
      });
    }

    /**
     * Navigates to a specific route.
     * @param {string} path - The route path to navigate to
     */
    navigateTo(path) {
      if (path === this.currentPath) {
        return;
      }
      
      this.currentPath = path;
      
      // Update browser history
      window.history.pushState(null, '', path);
      
      // Find matching route
      let componentFactory = null;
      let matchedPath = null;
      
      // First try exact match
      if (this.routes[path]) {
        componentFactory = this.routes[path];
        matchedPath = path;
      } else {
        // Try pattern matching (simple implementation)
        for (const [routePath, factory] of Object.entries(this.routes)) {
          if (routePath.includes(':') && this.pathMatches(routePath, path)) {
            componentFactory = factory;
            matchedPath = routePath;
            break;
          }
        }
        
        // Fallback to "/" or 404
        if (!componentFactory) {
          componentFactory = this.routes['/'] || ((p) => {
            const notFound = new jaffolding.Component('div');
            notFound.setText('404 - Page not found');
            return notFound;
          });
          matchedPath = '/';
        }
      }
      
      // Clear container
      while (this.container.firstChild) {
        this.container.removeChild(this.container.firstChild);
      }
      
      // Render new component
      const component = componentFactory(path);
      component.render(this.container);
    }

    /**
     * Gets the current path from the browser location.
     * @returns {string} - The current path
     * @private
     */
    getPath() {
      const path = window.location.pathname;
      return path || '/';
    }

    /**
     * Checks if a path matches a route pattern.
     * @param {string} routePath - The route pattern
     * @param {string} actualPath - The actual path
     * @returns {boolean} - Whether the path matches the pattern
     * @private
     */
    pathMatches(routePath, actualPath) {
      const routeParts = routePath.split('/');
      const actualParts = actualPath.split('/');
      
      if (routeParts.length !== actualParts.length) {
        return false;
      }
      
      for (let i = 0; i < routeParts.length; i++) {
        if (routeParts[i].startsWith(':')) {
          // This is a parameter, it matches anything
          continue;
        }
        
        if (routeParts[i] !== actualParts[i]) {
          return false;
        }
      }
      
      return true;
    }

    /**
     * Gets a parameter value from the current path.
     * @param {string} path - The current path
     * @param {string} paramName - The parameter name
     * @returns {string|null} - The parameter value or null if not found
     */
    getParameter(path, paramName) {
      for (const routePath of Object.keys(this.routes)) {
        if (routePath.includes(':') && this.pathMatches(routePath, path)) {
          const routeParts = routePath.split('/');
          const actualParts = path.split('/');
          
          for (let i = 0; i < routeParts.length; i++) {
            if (routeParts[i].startsWith(':') && 
                routeParts[i].substring(1) === paramName) {
              return actualParts[i];
            }
          }
        }
      }
      
      return null;
    }
  }

  // Export the Router class
  window.jaffolding = window.jaffolding || {};
  window.jaffolding.Router = Router;
})();

File: src/main/java/com/example/framework/Router.java
package com.example.framework;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

import org.teavm.jso.browser.Location;
import org.teavm.jso.browser.Window;
import org.teavm.jso.dom.html.HTMLElement;

/**
 * Simple client-side router for the framework.
 */
public class Router {
    private static final Window window = Window.current();
    private HTMLElement container;
    private Map<String, Function<String, Component>> routes = new HashMap<>();
    private String currentPath;
    
    public Router(HTMLElement container) {
        this.container = container;
        
        // Listen for popstate events (back/forward navigation)
        window.addEventListener("popstate", evt -> {
            navigateTo(getPath());
        });
    }
    
    public Router addRoute(String path, Function<String, Component> componentFactory) {
        routes.put(path, componentFactory);
        return this;
    }
    
    public void start() {
        // Handle initial route
        navigateTo(getPath());
        
        // Intercept link clicks for client-side routing
        window.addEventListener("click", evt -> {
            HTMLElement target = (HTMLElement) evt.getTarget();
            if ("A".equalsIgnoreCase(target.getTagName()) && 
                target.hasAttribute("href") && 
                !target.getAttribute("href").startsWith("http")) {
                
                evt.preventDefault();
                navigateTo(target.getAttribute("href"));
            }
        });
    }
    
    public void navigateTo(String path) {
        if (path.equals(currentPath)) {
            return;
        }
        
        currentPath = path;
        
        // Update browser history
        window.getHistory().pushState(null, "", path);
        
        // Find matching route
        Function<String, Component> componentFactory = null;
        String matchedPath = null;
        
        // First try exact match
        if (routes.containsKey(path)) {
            componentFactory = routes.get(path);
            matchedPath = path;
        } else {
            // Try pattern matching (simple implementation)
            for (Map.Entry<String, Function<String, Component>> entry : routes.entrySet()) {
                String routePath = entry.getKey();
                if (routePath.contains(":") && pathMatches(routePath, path)) {
                    componentFactory = entry.getValue();
                    matchedPath = routePath;
                    break;
                }
            }
            
            // Fallback to "/" or 404
            if (componentFactory == null) {
                componentFactory = routes.getOrDefault("/", p -> new Component("div")
                    .setText("404 - Page not found"));
                matchedPath = "/";
            }
        }
        
        // Clear container
        while (container.getFirstChild() != null) {
            container.removeChild(container.getFirstChild());
        }
        
        // Render new component
        Component component = componentFactory.apply(path);
        component.render(container);
    }
    
    private String getPath() {
        Location location = window.getLocation();
        String path = location.getPathname();
        return path.isEmpty() ? "/" : path;
    }
    
    private boolean pathMatches(String routePath, String actualPath) {
        String[] routeParts = routePath.split("/");
        String[] actualParts = actualPath.split("/");
        
        if (routeParts.length != actualParts.length) {
            return false;
        }
        
        for (int i = 0; i < routeParts.length; i++) {
            if (routeParts[i].startsWith(":")) {
                // This is a parameter, it matches anything
                continue;
            }
            
            if (!routeParts[i].equals(actualParts[i])) {
                return false;
            }
        }
        
        return true;
    }
    
    public String getParameter(String path, String paramName) {
        for (String routePath : routes.keySet()) {
            if (routePath.contains(":") && pathMatches(routePath, path)) {
                String[] routeParts = routePath.split("/");
                String[] actualParts = path.split("/");
                
                for (int i = 0; i < routeParts.length; i++) {
                    if (routeParts[i].startsWith(":") && 
                        routeParts[i].substring(1).equals(paramName)) {
                        return actualParts[i];
                    }
                }
            }
        }
        
        return null;
    }
}